File: frost-taproot/src/ecc/hash.rs
// Mirrors ref/frost/src/ecc/hash.ts
// FROST-secp256k1-SHA256-v1 hash functions H1–H5.
// Spec: draft-irtf-cfrg-frost-15, section 6.5.

use k256::{
    elliptic_curve::hash2curve::{ExpandMsg, ExpandMsgXmd, Expander},
    Scalar,
};
use sha2::{Digest, Sha256};

use super::util::mod_n;
use k256::U256;

const DOMAIN: &str = "FROST-secp256k1-SHA256-v1";

/// Build the DST for a given sub-tag.
fn dst(sub: &str) -> Vec<u8> {
    format!("{}{}", DOMAIN, sub).into_bytes()
}

/// hash_to_field using XMD with SHA-256, outputting one field element mod N.
///
/// Mirrors noble's `hash_to_field(msg, 1, { m:1, p:N, k:128, expand:'xmd', hash:sha256, DST })`:
///   1. expand_message_xmd → 48 uniform bytes
///   2. interpret as a big-endian 384-bit integer
///   3. reduce mod N
///
/// The 384-bit → mod-N reduction is done as:
///   value = hi_128_bits * 2^256 + lo_256_bits  (mod N)
fn hash_to_field_n(msg: &[u8], dst_bytes: &[u8]) -> Scalar {
    // Step 1: expand_message_xmd → 48 bytes
    let mut uniform = [0u8; 48];
    ExpandMsgXmd::<Sha256>::expand_message(&[msg], &[dst_bytes], 48)
        .expect("expand_message failed")
        .fill_bytes(&mut uniform);

    // Step 2 & 3: interpret 48 bytes as big-endian integer, reduce mod N.
    // Split into high 16 bytes and low 32 bytes.
    // value = hi * 2^256 + lo  (mod N)
    let (hi16, lo32_slice) = uniform.split_at(16);

    let mut lo32 = [0u8; 32];
    lo32.copy_from_slice(lo32_slice);
    let lo_scalar = mod_n(U256::from_be_slice(&lo32));

    // 2^256 mod N: since 2^256 = N + (2^256 - N), we compute it as a scalar.
    // 2^256 - N = 0x14551231950b75fc4402da1732fc9bebf
    // Represented as 32 bytes (fits in 17 bytes):
    let two256_mod_n = {
        // 2^256 mod N = 2^256 - N (since 2^256 > N)
        // = 0x014551231950b75fc4402da1732fc9bebf (17 bytes)
        // As 32-byte big-endian:
        let bytes =
            hex_literal_32("000000000000000000000000000000014551231950b75fc4402da1732fc9bebf");
        mod_n(U256::from_be_slice(&bytes))
    };

    // hi as a scalar (16 bytes → pad to 32)
    let mut hi32 = [0u8; 32];
    hi32[16..].copy_from_slice(hi16);
    let hi_scalar = mod_n(U256::from_be_slice(&hi32));

    hi_scalar * two256_mod_n + lo_scalar
}

/// Parse a hex literal into a 32-byte array (must be exactly 64 hex chars).
fn hex_literal_32(s: &str) -> [u8; 32] {
    // Pad or truncate to 64 hex chars (32 bytes), right-aligned.
    let padded = format!("{:0>64}", s);
    let b = (0..32)
        .map(|i| u8::from_str_radix(&padded[i * 2..i * 2 + 2], 16).unwrap())
        .collect::<Vec<_>>();
    b.try_into().unwrap()
}

/// H1: rho — binding factor hash.
pub fn h1(msg: &[u8]) -> [u8; 32] {
    let s = hash_to_field_n(msg, &dst("rho"));
    s.to_bytes().into()
}

/// H2: chal — challenge hash.
pub fn h2(msg: &[u8]) -> [u8; 32] {
    let s = hash_to_field_n(msg, &dst("chal"));
    s.to_bytes().into()
}

/// H3: nonce — nonce generation hash.
pub fn h3(msg: &[u8]) -> [u8; 32] {
    let s = hash_to_field_n(msg, &dst("nonce"));
    s.to_bytes().into()
}

/// H4: msg — plain SHA-256 with domain prefix.
pub fn h4(msg: &[u8]) -> [u8; 32] {
    let prefix = dst("msg");
    let mut hasher = Sha256::new();
    hasher.update(&prefix);
    hasher.update(msg);
    hasher.finalize().into()
}

/// H5: com — plain SHA-256 with domain prefix.
pub fn h5(msg: &[u8]) -> [u8; 32] {
    let prefix = dst("com");
    let mut hasher = Sha256::new();
    hasher.update(&prefix);
    hasher.update(msg);
    hasher.finalize().into()
}

File: frost-taproot-go/ecc/hash.go
// Package ecc provides elliptic curve operations for secp256k1.
package ecc

import (
	"crypto/sha256"
	"encoding/binary"
	"math/big"
)

const domain = "FROST-secp256k1-SHA256-v1"

// dst builds the domain separation tag.
func dst(sub string) []byte {
	return []byte(domain + sub)
}

// hashToFieldN hashes a message to a scalar mod N using XMD.
func hashToFieldN(msg, dstBytes []byte) *big.Int {
	// expand_message_xmd to 48 bytes using SHA-256
	uniform := expandMsgXmd(msg, dstBytes, 48)

	// Split into high 16 bytes and low 32 bytes
	hi16 := uniform[:16]
	lo32 := uniform[16:]

	// lo_scalar
	var loArr [32]byte
	copy(loArr[:], lo32)
	loScalar := ScalarFromBytes(loArr)

	// 2^256 mod N
	two256ModN, _ := new(big.Int).SetString("014551231950B75FC4402DA1732FC9BEBF", 16)

	// hi_scalar (16 bytes padded to 32)
	var hiArr [32]byte
	copy(hiArr[16:], hi16)
	hiScalar := ScalarFromBytes(hiArr)

	// result = hi * 2^256 + lo mod N
	result := ScalarMul(hiScalar, two256ModN)
	result = ScalarAdd(result, loScalar)
	return result
}

// expandMsgXmd implements expand_message_xmd with SHA-256.
func expandMsgXmd(msg, dstBytes []byte, lenInBytes int) []byte {
	b := 32 // SHA-256 block size
	ell := (lenInBytes + b - 1) / b

	// DST_prime = DST || I2OSP(len(DST), 1)
	dstPrime := make([]byte, len(dstBytes)+1)
	copy(dstPrime, dstBytes)
	dstPrime[len(dstBytes)] = byte(len(dstBytes))

	// Z_pad = I2OSP(0, s)
	zPad := make([]byte, 64) // s = 64 for SHA-256

	// l_i_b_str = I2OSP(len_in_bytes, 2)
	lenBytes := make([]byte, 2)
	binary.BigEndian.PutUint16(lenBytes, uint16(lenInBytes))

	// msg_prime = Z_pad || msg || l_i_b_str || I2OSP(0, 1) || DST_prime
	msgPrime := make([]byte, 0, len(zPad)+len(msg)+2+1+len(dstPrime))
	msgPrime = append(msgPrime, zPad...)
	msgPrime = append(msgPrime, msg...)
	msgPrime = append(msgPrime, lenBytes...)
	msgPrime = append(msgPrime, 0)
	msgPrime = append(msgPrime, dstPrime...)

	// b_0 = H(msg_prime)
	b0 := sha256.Sum256(msgPrime)

	// b_1 = H(b_0 || I2OSP(1, 1) || DST_prime)
	h1Input := make([]byte, 0, 32+1+len(dstPrime))
	h1Input = append(h1Input, b0[:]...)
	h1Input = append(h1Input, 1)
	h1Input = append(h1Input, dstPrime...)
	b1 := sha256.Sum256(h1Input)

	// uniform_bytes = b_1
	uniform := make([]byte, 0, lenInBytes)
	uniform = append(uniform, b1[:]...)

	// For i = 2 to ell:
	for i := 2; i <= ell; i++ {
		// b_i = H(strxor(b_0, b_(i-1)) || I2OSP(i, 1) || DST_prime)
		xor := make([]byte, 32)
		for j := range xor {
			xor[j] = b0[j] ^ uniform[len(uniform)-32+j]
		}
		hiInput := make([]byte, 0, 32+1+len(dstPrime))
		hiInput = append(hiInput, xor...)
		hiInput = append(hiInput, byte(i))
		hiInput = append(hiInput, dstPrime...)
		bi := sha256.Sum256(hiInput)
		uniform = append(uniform, bi[:]...)
	}

	return uniform[:lenInBytes]
}

// H1 computes the binding factor hash (rho).
func H1(msg []byte) [32]byte {
	s := hashToFieldN(msg, dst("rho"))
	return ScalarToBytes(s)
}

// H2 computes the challenge hash (chal).
func H2(msg []byte) [32]byte {
	s := hashToFieldN(msg, dst("chal"))
	return ScalarToBytes(s)
}

// H3 computes the nonce generation hash.
func H3(msg []byte) [32]byte {
	s := hashToFieldN(msg, dst("nonce"))
	return ScalarToBytes(s)
}

// H4 computes the message hash.
func H4(msg []byte) [32]byte {
	prefix := dst("msg")
	h := sha256.New()
	h.Write(prefix)
	h.Write(msg)
	var out [32]byte
	h.Sum(out[:0])
	return out
}

// H5 computes the commitment hash.
func H5(msg []byte) [32]byte {
	prefix := dst("com")
	h := sha256.New()
	h.Write(prefix)
	h.Write(msg)
	var out [32]byte
	h.Sum(out[:0])
	return out
}

// TagHash computes a BIP340-style tagged hash prefix.
func TagHash(tag string) [64]byte {
	h := sha256.Sum256([]byte(tag))
	var out [64]byte
	copy(out[:32], h[:])
	copy(out[32:], h[:])
	return out
}

// Hash340 computes a BIP340 tagged hash.
func Hash340(tag string, data [][]byte) [32]byte {
	prefix := TagHash(tag)
	h := sha256.New()
	h.Write(prefix[:])
	for _, d := range data {
		h.Write(d)
	}
	var out [32]byte
	h.Sum(out[:0])
	return out
}
